File: docs/LearnAlgorithm/LeetCode/Problems/97.Interleaving-String/solution0.py
class Solution0:
    """
    Timeout
    """
    def isInterleave(self, s1: str, s2: str, s3: str) -> bool:
        """
        >>> sol = Solution0()
        >>> sol.isInterleave("abbbbbbcabbacaacccababaabcccabcacbcaabbbacccaaaaaababbbacbb", "ccaacabbacaccacababbbbabbcacccacccccaabaababacbbacabbbbabc", "cacbabbacbbbabcbaacbbaccacaacaacccabababbbababcccbabcabbaccabcccacccaabbcbcaccccaaaaabaaaaababbbbacbbabacbbacabbbbabc")
        True
        >>> sol.isInterleave("aabcc", "dbbca", "aadbbbaccc")
        False
        >>> sol.isInterleave("aabcc", "dbbca", "aadbbcbcac")
        True
        """
        sum_all = len(s1) + len(s2) + len(s3)
        if sum_all == 0:
            return True
        elif len(s3) == 0 and sum_all != 0:
            return False
        else:
            first_match, second_match = len(s1) and s1[0] == s3[0], len(s2) and s2[0] == s3[0]
            first_success = self.isInterleave(s1[1:], s2[0:], s3[1:]) if first_match else False
            second_success = self.isInterleave(s1[0:], s2[1:], s3[1:]) if second_match else False
            return first_success or second_success


if __name__ == '__main__':
    import doctest
    doctest.testmod()

File: docs/LearnAlgorithm/LeetCode/Problems/97.Interleaving-String/solution0.cpp
#include <iostream>
#include <cassert>
using namespace std;

class Solution0 {
public:
    bool isInterleave(string s1, string s2, string s3) {
        unsigned long sum_all = s1.length() + s2.length() + s3.length();
        if (sum_all == 0) {
            return true;
        } else if (sum_all != 0 && s3.length() == 0) {
            return false;
        } else {
            bool first_match = s1.length() && s1[0] == s3[0];
            bool second_match = s2.length() && s2[0] == s3[0];
            bool first_success = first_match ? this->isInterleave(s1.substr(1), s2, s3.substr(1)) : false;
            bool second_success = second_match ? this->isInterleave(s1, s2.substr(1), s3.substr(1)) : false;
            return first_success || second_success;
        }
    }
};

int main() {
    Solution0 sol;
    bool res = sol.isInterleave("aabcc", "dbbca", "aadbbbaccc");
    assert(res == 0);
    res = sol.isInterleave("abbbbbbcabbacaacccababaabcccabcacbcaabbbacccaaaaaababbbacbb", "ccaacabbacaccacababbbbabbcacccacccccaabaababacbbacabbbbabc", "cacbabbacbbbabcbaacbbaccacaacaacccabababbbababcccbabcabbaccabcccacccaabbcbcaccccaaaaabaaaaababbbbacbbabacbbacabbbbabc");
    assert(res == 1);
    res = sol.isInterleave("aabcc", "dbbca", "aadbbcbcac");
    assert(res == 1);
    return 0;
}